kcwitools/optimal_extract_utility.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import print_function
+
+import numpy as np
+import matplotlib.pyplot as plt
+from numpy.polynomial.chebyshev import chebfit,chebval
+
+
+
+# Adopted from some online stuff that uses these for optimal extraction
+
+def wavelegth_fitter( fluxcube, varcube, aperture_mask, wavelength, order, aperture_indices, plot=False, verbose=False):
+
+	"""
+	Fit polynomials in the wavelength direction of each pixel in an aperture. 
+	
+	Input: 
+		 fluxcube 			:    KCWI flux cube 
+		 varceub  			:	 KCWI variance cube
+		 aperture_mask  	:    3D mask for the pseudo-slit. Good values=True, Bad ones =False
+		 wavelength     	:    wavelength grid of the fluxcube
+		 order          	:    Chebyshev polynomial order for fitting
+		 aperture_indices 	:    Indices of the cube corresponding to pseudo-slit aperture, in an Nx2 array. 
+
+		 plot [False]		:  If true plots the polynomials (both before and after normalising spatially)    
+		 verbose [False]    :  True prints stuff in terminal
+
+
+	"""
+
+
+	wavedim,ydim,xdim=fluxcube.shape
+	wavelength_fits=np.zeros([wavedim, ydim, xdim])
+
+
+	#Lamdas must be scaled to between -1 and 1 for Chebyshev fitting
+	min_val=wavelength.min()
+	max_val=wavelength.max()
+	scaled_l=np.array([2*(lamda-min_val)/(max_val-min_val) -1 for lamda in wavelength])
+
+	for n, (i, j) in enumerate(zip(aperture_indices[0], aperture_indices[1])):
+
+		"""Take each spaxel and fit a polynomial along the wavelength direction, each pixel weighted by one over its variance"""
+		if verbose==True:
+			print("Fitting Pixel {}".format(n))
+
+		weights=1.0/varcube[:, i, j]
+
+		#Fit the Chebyshev coefficients
+		coefficients=chebfit(scaled_l, fluxcube[:, i, j], order, w=weights)
+
+		#Make a polynomial from these
+		polynomial=chebval(scaled_l, coefficients)
+		#Ensure all values are positve
+		polynomial[polynomial<0]=0
+
+
+		wavelength_fits[:, i, j]=polynomial
+
+	if plot==True:
+		for i, j in zip(aperture_indices[0], aperture_indices[1]):
+			plt.plot(wavelength, wavelength_fits[:, :])
+
+		plt.show()
+
+	if verbose==True:
+		print("Normalising Spatially")
+	#Normalise spatially
+	spatial_norm=np.sum(np.sum(wavelength_fits, axis=2), axis=1)
+
+	# This is the P(x, lambda) value from Horne 1986 in 3D
+	mcube=wavelength_fits/spatial_norm[:, np.newaxis, np.newaxis]
+
+
+
+	if plot==True:
+		for i, j in zip(aperture_indices[0], aperture_indices[1]):
+
+			plt.plot(wavelength, mcube[:,i, j])
+
+		plt.show()
+
+	return mcube
+
+
+
+def extract1d(datacube,verbose=False):
```

Comment:
Add doc strings for this and the methods that follow.

Reply:
This is depreciated.